### PR TITLE
FIX: Gitops Config Validation

### PR DESCRIPTION
**Pull request by kartik-579**
# Description

In the current support for Gitops configuration, the validation for the same is not implemented when it is being provided by the user. Added validation at the step when a configuration is saved/updated.

# How Has This Been Tested?
The feature is tested on localhost as well as a demo environment for all the 3 providers, i.e. GitHub, GitLab, Azure.

### CODE REVIEW

File: internal/util/GitService.go [hunk 1]
```diff
@@ -68,6 +75,43 @@ func (factory *GitFactory) Reload() error {
 	return nil
 }
 
+type GitOpsConfigDtoTemp struct {
```
Comment (vikramdevtron) [inline internal/util/GitService.go:78]: redundant GitOpsConfigDtoTemp, move it to common DTO.

File: internal/util/GitService.go [hunk 2]
```diff
@@ -206,45 +250,61 @@ func NewGitLabClient(config *GitConfig, logger *zap.SugaredLogger, gitService Gi
 		return nil, nil
 	}
 }
-
-func (impl GitLabClient) CreateRepository(name, description string) (url string, isNew bool, err error) {
+func (impl GitLabClient) DeleteRepository(name, userName, gitHubOrgName, azureProjectName string) error {
+	err := impl.DeleteProject(name)
```
Comment (vikramdevtron) [inline internal/util/GitService.go:254]: log error

File: internal/util/GitService.go [hunk 3]
```diff
@@ -206,45 +250,61 @@ func NewGitLabClient(config *GitConfig, logger *zap.SugaredLogger, gitService Gi
...
+	return err
+}
+func (impl GitLabClient) CreateRepository(name, description string) (url string, isNew bool, detailedError DetailedError) {
+	detailedError.StageErrorMap = make(map[string]error)
 	impl.logger.Debugw("gitlab app create request ", "name", name, "description", description)
 	repoUrl, err := impl.GetRepoUrl(name)
 	if err != nil {
 		impl.logger.Errorw("error in getting repo url ", "project", name, "err", err)
-		return "", false, err
+		detailedError.StageErrorMap["GetRepoUrl"] = err
```
Comment (vikramdevtron) [inline internal/util/GitService.go:263]: use Constant for Key

File: internal/util/GitService.go [hunk 4]
```diff
@@ -206,45 +250,61 @@ func NewGitLabClient(config *GitConfig, logger *zap.SugaredLogger, gitService Gi
...
+			return "", true, detailedError
 		}
+		detailedError.SuccessfulStages = append(detailedError.SuccessfulStages, "CreateRepo")
 	}
 	repoUrl = url
 	validated, err := impl.ensureProjectAvailability(name)
 	if err != nil {
 		impl.logger.Errorw("error in ensuring project availability ", "project", name, "err", err)
-		return "", true, err
+		detailedError.StageErrorMap["CLoneHttp"] = err
```
Comment (vikramdevtron) [inline internal/util/GitService.go:281]: use Constant for Key

File: internal/util/GitService.go [hunk 5]
```diff
@@ -206,45 +250,61 @@ func NewGitLabClient(config *GitConfig, logger *zap.SugaredLogger, gitService Gi
...
+		detailedError.StageErrorMap["CloneSsh"] = err
+		return "", true, detailedError
 	}
 	if !validated {
-		return "", true, fmt.Errorf("unable to validate project:%s  in given time", name)
+		detailedError.StageErrorMap["CloneSsh"] = fmt.Errorf("unable to validate project:%s in given time", name)
+		return "", true, detailedError
 	}
-	return url, true, nil
+	detailedError.SuccessfulStages = append(detailedError.SuccessfulStages, "CloneSsh")
```
Comment (vikramdevtron) [inline internal/util/GitService.go:306]: add "CloneSsh" to Constant , or other similar items

File: internal/sql/repository/GitOpsConfigRepository.go [hunk 6]
```diff
@@ -97,3 +103,28 @@ func (impl *GitOpsConfigRepositoryImpl) GetGitOpsConfigActive() (*GitOpsConfig,
 	err := impl.dbConnection.Model(&model).Where("active = ?", true).Limit(1).Select()
 	return &model, err
 }
+
+//--------------gitOps-validation-status--------------------
+
+type GitOpsConfigValidationStatus struct {
+	tableName        struct{}  `sql:"gitops_config_validation_status" pg:",discard_unknown_columns"`
+	Id               int       `sql:"id,pk"`
+	Provider         string    `sql:"provider"`
```
[the remark this answers is not in the document]
Comment (kartik-579) [reply internal/sql/repository/GitOpsConfigRepository.go:112]: not needed as this table is only used for storing the latest validation status and not for every config

**Review by nishant-d — CHANGES_REQUESTED** [review]

**Review by nishant-d — APPROVED** [review]

**Review by nishant-d — APPROVED** [review]

**Review by nishant-d — APPROVED** [review]

**Pull request merged**